File: templates/dart/protocols/mavlink_session.dart
import 'dart:async';
import 'dart:typed_data';

import '../mavlink_dialect.dart';
import '../mavlink_frame.dart';
import '../mavlink_message.dart';
import '../mavlink_parser.dart';
import '../mavlink_version.dart';
import 'mavlink_cancellation.dart';
import 'mavlink_link.dart';

/// Thrown when an expected MAVLink message is not received in time.
class MavlinkTimeoutException implements Exception {
  MavlinkTimeoutException(this.message, this.timeout);

  final String message;
  final Duration timeout;

  @override
  String toString() => 'MavlinkTimeoutException: $message (timeout: $timeout)';
}

/// Handle returned by [MavlinkSession.listenMessage]; call [cancel] to unsubscribe.
class MavlinkMessageSubscription {
  MavlinkMessageSubscription._(this._cancel);

  final void Function() _cancel;
  var _active = true;

  bool get isActive => _active;

  void cancel() {
    if (!_active) {
      return;
    }
    _active = false;
    _cancel();
  }
}

class _PendingFrameWait {
  _PendingFrameWait({
    required this.predicate,
    required this.completer,
    required this.timer,
    this.cancel,
    this.cancelSub,
  });

  final bool Function(MavlinkFrame frame) predicate;
  final Completer<MavlinkFrame> completer;
  final Timer timer;
  final MavlinkCancellationToken? cancel;
  StreamSubscription<void>? cancelSub;
}

/// Framing, sequencing, and message dispatch over a [MavlinkLink].
///
/// Protocol implementations use a session to send typed messages and wait for
/// responses without knowing whether the link is USB, UDP, or in-memory.
class MavlinkSession {
  MavlinkSession({
    required MavlinkDialect dialect,
    required MavlinkLink link,
    required this.systemId,
    required this.componentId,
    this.version = MavlinkVersion.v2,
  })  : _dialect = dialect,
        _link = link {
    _parser = MavlinkParser(_dialect);
    _parser.stream.listen(_onFrame);
    _subscription = _link.receive.listen(_parser.parse);
  }

  final MavlinkDialect _dialect;
  final MavlinkLink _link;
  final int systemId;
  final int componentId;
  final MavlinkVersion version;

  late final MavlinkParser _parser;
  late final StreamSubscription<Uint8List> _subscription;
  final _framesController = StreamController<MavlinkFrame>.broadcast();
  final List<_PendingFrameWait> _pendingWaits = [];
  final List<MavlinkFrame> _recentFrames = [];
  static const _recentFrameCapacity = 64;
  int _sequence = 0;
  var _closed = false;

  MavlinkDialect get dialect => _dialect;

  /// All frames parsed from the link (before filtering).
  Stream<MavlinkFrame> get frames => _framesController.stream;

  /// Typed message stream filtered by [fromSystemId] / [fromComponentId].
  Stream<T> onMessage<T extends MavlinkMessage>({int? fromSystemId, int? fromComponentId}) {
    return frames
        .where((frame) {
          if (fromSystemId != null && frame.systemId != fromSystemId) {
            return false;
          }
          if (fromComponentId != null && frame.componentId != fromComponentId) {
            return false;
          }
          return frame.message is T;
        })
        .map((frame) => frame.message as T);
  }

  /// Message stream filtered by MAVLink message id.
  Stream<MavlinkMessage> subscribeMessageId(int messageId, {int? fromSystemId, int? fromComponentId}) {
    return frames
        .where((frame) {
          if (frame.message.mavlinkMessageId != messageId) {
            return false;
          }
          if (fromSystemId != null && frame.systemId != fromSystemId) {
            return false;
          }
          if (fromComponentId != null && frame.componentId != fromComponentId) {
            return false;
          }
          return true;
        })
        .map((frame) => frame.message);
  }

  /// Register a callback for messages of type [T]. Returns a subscription handle.
  MavlinkMessageSubscription listenMessage<T extends MavlinkMessage>(
    void Function(T message, MavlinkFrame frame) onData, {
    int? fromSystemId,
    int? fromComponentId,
  }) {
    final subscription = frames.listen((frame) {
      if (fromSystemId != null && frame.systemId != fromSystemId) {
        return;
      }
      if (fromComponentId != null && frame.componentId != fromComponentId) {
        return;
      }
      if (frame.message is T) {
        onData(frame.message as T, frame);
      }
    });

    return MavlinkMessageSubscription._(() => subscription.cancel());
  }

  /// Send a typed MAVLink message as a framed packet.
  Future<void> send(MavlinkMessage message) async {
    if (_closed) {
      throw StateError('MavlinkSession is closed');
    }

    final frame = version == MavlinkVersion.v2
        ? MavlinkFrame.v2(_sequence++ & 0xff, systemId, componentId, message)
        : MavlinkFrame.v1(_sequence++ & 0xff, systemId, componentId, message);

    await _link.send(frame.serialize());
  }

  /// Wait for the first frame matching [predicate].
  Future<MavlinkFrame> waitForFrame({
    required bool Function(MavlinkFrame frame) predicate,
    Duration timeout = const Duration(seconds: 5),
    MavlinkCancellationToken? cancel,
  }) {
    cancel?.throwIfCancelled();

    final completer = Completer<MavlinkFrame>();
    late final _PendingFrameWait wait;
    wait = _PendingFrameWait(
      predicate: predicate,
      completer: completer,
      timer: Timer(timeout, () {
        _pendingWaits.remove(wait);
        if (!completer.isCompleted) {
          completer.completeError(
            MavlinkTimeoutException('Timed out waiting for frame', timeout),
          );
        }
      }),
      cancel: cancel,
    );

    if (cancel != null) {
      if (cancel.isCancelled) {
        wait.timer.cancel();
        return Future.error(MavlinkCancelledException());
      }
      wait.cancelSub = cancel.onCancel.listen((_) {
        _pendingWaits.remove(wait);
        wait.timer.cancel();
        if (!completer.isCompleted) {
          completer.completeError(MavlinkCancelledException());
        }
      });
    }

    _pendingWaits.add(wait);

    for (final frame in List<MavlinkFrame>.from(_recentFrames)) {
      if (!predicate(frame)) {
        continue;
      }
      _recentFrames.remove(frame);
      wait.timer.cancel();
      wait.cancelSub?.cancel();
      _pendingWaits.remove(wait);
      if (!completer.isCompleted) {
        completer.complete(frame);
      }
      return completer.future;
    }

    completer.future.whenComplete(() {
      wait.timer.cancel();
      wait.cancelSub?.cancel();
      _pendingWaits.remove(wait);
    });

    return completer.future;
  }

  /// Wait for the first message matching [predicate].
  Future<MavlinkMessage> waitForMessage({
    required bool Function(MavlinkMessage message) predicate,
    int? fromSystemId,
    int? fromComponentId,
    Duration timeout = const Duration(seconds: 5),
    MavlinkCancellationToken? cancel,
  }) async {
    final frame = await waitForFrame(
      predicate: (frame) {
        if (fromSystemId != null && frame.systemId != fromSystemId) {
          return false;
        }
        if (fromComponentId != null && frame.componentId != fromComponentId) {
          return false;
        }
        return predicate(frame.message);
      },
      timeout: timeout,
      cancel: cancel,
    );
    return frame.message;
  }

  /// Wait for the first message of type [T].
  Future<T> waitForMessageType<T extends MavlinkMessage>({
    int? fromSystemId,
    int? fromComponentId,
    Duration timeout = const Duration(seconds: 5),
    MavlinkCancellationToken? cancel,
  }) async {
    final message = await waitForMessage(
      predicate: (message) => message is T,
      fromSystemId: fromSystemId,
      fromComponentId: fromComponentId,
      timeout: timeout,
      cancel: cancel,
    );
    return message as T;
  }

  void _onFrame(MavlinkFrame frame) {
    if (_closed) {
      return;
    }

    _framesController.add(frame);
    _recentFrames.add(frame);
    if (_recentFrames.length > _recentFrameCapacity) {
      _recentFrames.removeAt(0);
    }

    for (final wait in List<_PendingFrameWait>.from(_pendingWaits)) {
      if (!wait.predicate(frame)) {
        continue;
      }

      wait.timer.cancel();
      wait.cancelSub?.cancel();
      _pendingWaits.remove(wait);
      _recentFrames.remove(frame);
      if (!wait.completer.isCompleted) {
        wait.completer.complete(frame);
      }
      break;
    }
  }

  Future<void> close() async {
    if (_closed) {
      return;
    }
    _closed = true;

    for (final wait in List<_PendingFrameWait>.from(_pendingWaits)) {
      wait.timer.cancel();
      wait.cancelSub?.cancel();
      if (!wait.completer.isCompleted) {
        wait.completer.completeError(StateError('MavlinkSession is closed'));
      }
    }
    _pendingWaits.clear();

    await _subscription.cancel();
    await _framesController.close();
    await _link.close();
  }
}

File: templates/csharp/protocols/mavlink_session.cs
using System.Runtime.CompilerServices;
using System.Threading.Channels;
using Mavlink.Dialects;

namespace Mavlink;

/// <summary>Thrown when an expected MAVLink message is not received in time.</summary>
public sealed class MavlinkTimeoutException : Exception
{
    public MavlinkTimeoutException(string message, TimeSpan timeout)
        : base($"{message} (timeout: {timeout})")
    {
        MessageText = message;
        Timeout = timeout;
    }

    public string MessageText { get; }

    public TimeSpan Timeout { get; }
}

/// <summary>
/// Handle returned by <see cref="MavlinkSession.ListenMessage{T}"/>; call <see cref="Cancel"/> to unsubscribe.
/// </summary>
public sealed class MavlinkMessageSubscription
{
    private readonly Action _cancel;
    private bool _active = true;

    internal MavlinkMessageSubscription(Action cancel) => _cancel = cancel;

    public bool IsActive => _active;

    public void Cancel()
    {
        if (!_active)
        {
            return;
        }

        _active = false;
        _cancel();
    }
}

/// <summary>
/// Framing, sequencing, and message dispatch over a <see cref="MavlinkLink"/>.
/// </summary>
public sealed class MavlinkSession : IAsyncDisposable
{
    private const int RecentFrameCapacity = 64;

    private readonly MavlinkDialect _dialect;
    private readonly MavlinkLink _link;
    private readonly MavlinkParser _parser;
    private readonly Channel<MavlinkFrame> _frames = Channel.CreateUnbounded<MavlinkFrame>();
    private readonly List<PendingFrameWait> _pendingWaits = new();
    private readonly List<MavlinkFrame> _recentFrames = new();
    private readonly CancellationTokenSource _cts = new();
    private readonly Task _receiveTask;

    private int _parsedFrameCount;
    private int _sequence;
    private bool _closed;

    public MavlinkSession(
        MavlinkDialect dialect,
        MavlinkLink link,
        byte systemId,
        byte componentId,
        MavlinkVersion version = MavlinkVersion.V2)
    {
        _dialect = dialect;
        _link = link;
        SystemId = systemId;
        ComponentId = componentId;
        Version = version;
        _parser = new MavlinkParser(dialect);
        _receiveTask = Task.Run(ProcessReceiveAsync);
    }

    public MavlinkDialect Dialect => _dialect;

    public byte SystemId { get; }

    public byte ComponentId { get; }

    public MavlinkVersion Version { get; }

    /// <summary>All frames parsed from the link (before filtering).</summary>
    public IAsyncEnumerable<MavlinkFrame> Frames => _frames.Reader.ReadAllAsync();

    /// <summary>Typed message stream filtered by <paramref name="fromSystemId"/> / <paramref name="fromComponentId"/>.</summary>
    public async IAsyncEnumerable<T> OnMessage<T>(
        byte? fromSystemId = null,
        byte? fromComponentId = null,
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
        where T : MavlinkMessage
    {
        await foreach (var frame in Frames.WithCancellation(cancellationToken).ConfigureAwait(false))
        {
            if (fromSystemId is not null && frame.SystemId != fromSystemId.Value)
            {
                continue;
            }

            if (fromComponentId is not null && frame.ComponentId != fromComponentId.Value)
            {
                continue;
            }

            if (frame.Message is T message)
            {
                yield return message;
            }
        }
    }

    /// <summary>Message stream filtered by MAVLink message id.</summary>
    public async IAsyncEnumerable<MavlinkMessage> SubscribeMessageId(
        int messageId,
        byte? fromSystemId = null,
        byte? fromComponentId = null,
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        await foreach (var frame in Frames.WithCancellation(cancellationToken).ConfigureAwait(false))
        {
            if (frame.Message.MavlinkMessageId != messageId)
            {
                continue;
            }

            if (fromSystemId is not null && frame.SystemId != fromSystemId.Value)
            {
                continue;
            }

            if (fromComponentId is not null && frame.ComponentId != fromComponentId.Value)
            {
                continue;
            }

            yield return frame.Message;
        }
    }

    /// <summary>Register a callback for messages of type <typeparamref name="T"/>.</summary>
    public MavlinkMessageSubscription ListenMessage<T>(
        Action<T, MavlinkFrame> onData,
        byte? fromSystemId = null,
        byte? fromComponentId = null)
        where T : MavlinkMessage
    {
        var cts = new CancellationTokenSource();
        var task = Task.Run(async () =>
        {
            await foreach (var frame in Frames.WithCancellation(cts.Token).ConfigureAwait(false))
            {
                if (fromSystemId is not null && frame.SystemId != fromSystemId.Value)
                {
                    continue;
                }

                if (fromComponentId is not null && frame.ComponentId != fromComponentId.Value)
                {
                    continue;
                }

                if (frame.Message is T message)
                {
                    onData(message, frame);
                }
            }
        }, cts.Token);

        return new MavlinkMessageSubscription(() =>
        {
            cts.Cancel();
            cts.Dispose();
            _ = task;
        });
    }

    /// <summary>Send a typed MAVLink message as a framed packet.</summary>
    public async Task SendAsync(MavlinkMessage message, CancellationToken cancellationToken = default)
    {
        if (_closed)
        {
            throw new InvalidOperationException("MavlinkSession is closed");
        }

        var frame = Version == MavlinkVersion.V2
            ? MavlinkFrame.V2((byte)(_sequence++ & 0xFF), SystemId, ComponentId, message)
            : MavlinkFrame.V1((byte)(_sequence++ & 0xFF), SystemId, ComponentId, message);

        await _link.SendAsync(frame.Serialize(), cancellationToken).ConfigureAwait(false);
    }

    /// <summary>Wait for the first frame matching <paramref name="predicate"/>.</summary>
    public Task<MavlinkFrame> WaitForFrameAsync(
        Func<MavlinkFrame, bool> predicate,
        TimeSpan? timeout = null,
        MavlinkCancellationToken? cancel = null)
    {
        cancel?.ThrowIfCancelled();
        var effectiveTimeout = timeout ?? TimeSpan.FromSeconds(5);
        var tcs = new TaskCompletionSource<MavlinkFrame>(TaskCreationOptions.RunContinuationsAsynchronously);
        var wait = new PendingFrameWait(predicate, tcs, effectiveTimeout, cancel);
        RegisterWait(wait);

        foreach (var frame in _recentFrames.ToList())
        {
            if (!predicate(frame))
            {
                continue;
            }

            _recentFrames.Remove(frame);
            CompleteWait(wait);
            tcs.TrySetResult(frame);
            return tcs.Task;
        }

        return tcs.Task;
    }

    /// <summary>Wait for the first message matching <paramref name="predicate"/>.</summary>
    public async Task<MavlinkMessage> WaitForMessageAsync(
        Func<MavlinkMessage, bool> predicate,
        byte? fromSystemId = null,
        byte? fromComponentId = null,
        TimeSpan? timeout = null,
        MavlinkCancellationToken? cancel = null)
    {
        var frame = await WaitForFrameAsync(
            frame =>
            {
                if (fromSystemId is not null && frame.SystemId != fromSystemId.Value)
                {
                    return false;
                }

                if (fromComponentId is not null && frame.ComponentId != fromComponentId.Value)
                {
                    return false;
                }

                return predicate(frame.Message);
            },
            timeout,
            cancel).ConfigureAwait(false);

        return frame.Message;
    }

    /// <summary>Wait for the first message of type <typeparamref name="T"/>.</summary>
    public async Task<T> WaitForMessageTypeAsync<T>(
        byte? fromSystemId = null,
        byte? fromComponentId = null,
        TimeSpan? timeout = null,
        MavlinkCancellationToken? cancel = null)
        where T : MavlinkMessage
    {
        var message = await WaitForMessageAsync(
            message => message is T,
            fromSystemId,
            fromComponentId,
            timeout,
            cancel).ConfigureAwait(false);

        return (T)message;
    }

    public async ValueTask DisposeAsync()
    {
        await CloseAsync().ConfigureAwait(false);
    }

    public async Task CloseAsync()
    {
        if (_closed)
        {
            return;
        }

        _closed = true;
        _cts.Cancel();

        foreach (var wait in _pendingWaits.ToList())
        {
            CompleteWait(wait);
            wait.Completion.TrySetException(new InvalidOperationException("MavlinkSession is closed"));
        }

        _pendingWaits.Clear();
        _frames.Writer.TryComplete();

        try
        {
            await _receiveTask.ConfigureAwait(false);
        }
        catch (OperationCanceledException)
        {
        }

        await _link.CloseAsync().ConfigureAwait(false);
        _cts.Dispose();
    }

    private async Task ProcessReceiveAsync()
    {
        try
        {
            await foreach (var chunk in _link.Receive.WithCancellation(_cts.Token).ConfigureAwait(false))
            {
                _parser.Parse(chunk);
                while (_parsedFrameCount < _parser.Frames.Count)
                {
                    OnFrame(_parser.Frames[_parsedFrameCount++]);
                }
            }
        }
        catch (OperationCanceledException) when (_closed || _cts.IsCancellationRequested)
        {
        }
    }

    private void OnFrame(MavlinkFrame frame)
    {
        if (_closed)
        {
            return;
        }

        _frames.Writer.TryWrite(frame);
        _recentFrames.Add(frame);
        if (_recentFrames.Count > RecentFrameCapacity)
        {
            _recentFrames.RemoveAt(0);
        }

        foreach (var wait in _pendingWaits.ToList())
        {
            if (!wait.Predicate(frame))
            {
                continue;
            }

            _recentFrames.Remove(frame);
            CompleteWait(wait);
            wait.Completion.TrySetResult(frame);
            break;
        }
    }

    private void RegisterWait(PendingFrameWait wait)
    {
        _pendingWaits.Add(wait);
        wait.Timer = new Timer(
            _ =>
            {
                if (_pendingWaits.Remove(wait))
                {
                    wait.Completion.TrySetException(
                        new MavlinkTimeoutException("Timed out waiting for frame", wait.Timeout));
                    wait.DisposeRegistration();
                }
            },
            null,
            wait.Timeout,
            Timeout.InfiniteTimeSpan);

        if (wait.Cancel is not null)
        {
            wait.CancelRegistration = wait.Cancel.Register(() =>
            {
                if (_pendingWaits.Remove(wait))
                {
                    wait.Timer?.Dispose();
                    wait.Completion.TrySetException(new MavlinkCancelledException());
                    wait.DisposeRegistration();
                }
            });
        }

        _ = wait.Completion.Task.ContinueWith(
            _ =>
            {
                CompleteWait(wait);
                wait.DisposeRegistration();
            },
            TaskScheduler.Default);
    }

    private static void CompleteWait(PendingFrameWait wait)
    {
        wait.Timer?.Dispose();
        wait.Timer = null;
    }

    private sealed class PendingFrameWait
    {
        public PendingFrameWait(
            Func<MavlinkFrame, bool> predicate,
            TaskCompletionSource<MavlinkFrame> completion,
            TimeSpan timeout,
            MavlinkCancellationToken? cancel)
        {
            Predicate = predicate;
            Completion = completion;
            Timeout = timeout;
            Cancel = cancel;
        }

        public Func<MavlinkFrame, bool> Predicate { get; }

        public TaskCompletionSource<MavlinkFrame> Completion { get; }

        public TimeSpan Timeout { get; }

        public MavlinkCancellationToken? Cancel { get; }

        public Timer? Timer { get; set; }

        public IDisposable? CancelRegistration { get; set; }

        public void DisposeRegistration() => CancelRegistration?.Dispose();
    }
}
